### PR TITLE
Restore-DbaDatabase - Changed WhatIf message / added deep copy of FileList

### LINKED ISSUE #5036: Restore-DbaDatabase changes paths in the backup history object
### Before submitting a bug report:

- [x] Ensure you are able to reproduce it on the latest released version (_we release often_)
- [x] Verified this bug is not already reported in an issue
- [x] Verified errors are not related to permissions
- [x] Can reproduce in a clean PowerShell session (_clean = `powershell -NoProfile`_)
- [x] If issue is with `Copy-DbaDatabase`, using `Backup-DbaDatabase | Restore-DbaDatabase` still does not work

## Steps to Reproduce

```powershell
$historyObject = Get-DbaBackupHistory -SqlInstance lab-sds-ag1:7221 -Database Ag1Database1
$historyObject[0].FileList

FileType LogicalName      PhysicalName
-------- -----------      ------------
D        AG1Database1     C:\Data\MSSQL13.LIFE\AG1Database1.mdf
L        AG1Database1_log C:\Data\MSSQL13.LIFE\AG1Database1_log.ldf
D        AG1Database1_2   C:\Data\MSSQL13.LIFE\AG1Database1_2.ndf


$null = $historyObject | Restore-DbaDatabase -SqlInstance lab-sds-ag1:7221 -DatabaseName Ag1Database1_copy -TrustDbBackupHistory -ReplaceDbNameInFile -WithReplace -OutputScriptOnly
$historyObject[0].FileList

FileType LogicalName      PhysicalName
-------- -----------      ------------
D        AG1Database1     C:\Data\MSSQL13.LIFE\Ag1Database1_copy.mdf
L        AG1Database1_log C:\Data\MSSQL13.LIFE\Ag1Database1_copy_log.ldf
D        AG1Database1_2   C:\Data\MSSQL13.LIFE\Ag1Database1_copy_2.ndf


$null = $historyObject | Restore-DbaDatabase -SqlInstance lab-sds-ag1:7221 -DatabaseName Ag1Database1_copy -TrustDbBackupHistory -ReplaceDbNameInFile -WithReplace -OutputScriptOnly
$historyObject[0].FileList

FileType LogicalName      PhysicalName
-------- -----------      ------------
D        AG1Database1     C:\Data\MSSQL13.LIFE\Ag1Database1_copy_copy.mdf
L        AG1Database1_log C:\Data\MSSQL13.LIFE\Ag1Database1_copy_copy_log.ldf
D        AG1Database1_2   C:\Data\MSSQL13.LIFE\Ag1Database1_copy_copy_2.ndf

```

## Expected Behavior

$historyObject remaining unchanged

## Actual Behavior

Physical name gets replaced with the new database name every time the restore runs creating enormous filenames after a few runs:
Ag1Database1_copy_copy_copy_copy_copy_copy_copy_log.ldf

## Environmental data

 - PowerShell:
Name                           Value
----                           -----
PSVersion                      5.1.15063.1478
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.15063.1478
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

 - - SQL Server: 2016SP2

### LINKED ISSUE #6662: [Bug] Restore-DbaDatabase -Whatif returns a kind of empty result when it returns success
#### Environmental information

- Windows 10

<!--
🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
Please provide the output of the below script
```powershell
[pscustomobject]@{
>> 'PowerShell Version' = $PSVersionTable.PSVersion.ToString()
>> 'dbatools latest installed' = (Get-InstalledModule -Name dbatools).Version
>> 'Culture of OS' = (Get-Culture)
>> } | fl -force
```
-->

- PowerShell 5.1.18362.752
- dbatools 1      0      114    -1
- Culture: LCID 2055, de-CH, Switzerland

#### SQL Server:

Microsoft SQL Server 2019 (RTM-CU4) (KB4548597) - 15.0.4033.1 (X64) 	Mar 14 2020 16:10:35 	Copyright (C) 2019 Microsoft Corporation	Developer Edition (64-bit) on Windows 10 Pro 10.0 <X64> (Build 18363: )
Language: us_english

### Report

<!--
Things to consider:
- Errors received are not related to permissions?
- Have you tried the same command using powershell.exe instead of a hosted powershell instance like ISE or VS Code?
- If this refers to Copy-DbaDatabase can you replace the problem with Backup-DbaDatabase and Restore-DbaDatabase?
- Copy-DbaDatabase will not work in every environment and every situation. Instead, we try to ensure Backup & Restore work in your environment.
-->

The bug is not permission related. If -Whatif reports that the restore might be successful, it does return some information, but they are quite not helpful because they are empty. 


#### Host used

- [x] powershell.exe
- [x] ISE
- [x] VS Code
- [ ] Other (please specify)


#### Errors Received
none

<!--
🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
Provide the full error stack, you can obtain this by duplicating the error and then immediately running this command: `$error[0] | select *`
-->

#### Steps to Reproduce
Create a full backup and some *.trn files. The idea was to test the LSN chain before actual restore. 
```sql
Restore-DbaDatabase -SqlInstance $server_Name -Path $backup_Path -DatabaseName $db_name -WithReplace -EnableException -Whatif
```

#### Expected Behavior

<!--
What did you expect to happen?
-->

#### Actual Behavior
If successful, -whatif yields following message:
```
WhatIf: Ausführen des Vorgangs "
 Restore Database DatabaseName on ServerName
 from files:
 with these file moves:         
```                                                                                                                                                                                                                

from files and with these file moves are empty as are all other similar messages.

### PR DESCRIPTION
**Pull request by andreasjordan**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes #6662, fixes #5036)
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system

### CODE REVIEW

**Comment by andreasjordan** [comment]
image

**Comment by andreasjordan** [comment]
Not sure if the deep copy has some side effects. Will have to look at that.

**Comment by andreasjordan** [comment]
Ok, deep copy breaks Test-DbaLastBackup, more work has to be done...

**Comment by potatoqualitee** [comment]
Hey @Stuart-Moore, can you please review this PR? 🙏🏼

**Comment by potatoqualitee** [comment]
Will reach out to Stuart on Twitter, this command is ultra important so I want to ensure he gets a look.

**Comment by Stuart-Moore** [comment]
Does the copy not want to be a bit further into the function? All the ways of passing the backup info into the function end up in BackupHistory  objects, so should all be affected? Somewhere like line 619 in your modified file

Not played with this for a long time and have been rotting my brain with ADF and Dynamics so may have missed something.

The overall way of  doing it should work.

would be nice to see a test adding, just to make sure someone else doesn't  break it later on by mistake or by misunderstanding.

**Comment by andreasjordan** [comment]
Output of the test, if solution is not active:

image

**Comment by andreasjordan** [comment]
Hi @Stuart-Moore , I had another look at this and found a better solution. I realized that you do a lot of `$f = $f | Select-Object *,  ...` to get a new object with only that values from the piped in backup history. But as the FileList property is an array of objects, the new backup history object just references the same FileList objects. To implement a "deep copy" (is this really the correct word?) I just used the same technique as you did with the backup history object.
I added a test, you find a screenshot of the failing test a comment above.

File: functions/Restore-DbaDatabase.ps1 [hunk 1]
```diff
@@ -677,7 +679,7 @@ function Restore-DbaDatabase {
                 return
             }
             $pathSep = Get-DbaPathSep -Server $RestoreInstance
-            $null = $BackupHistory | Format-DbaBackupInformation -DataFileDirectory $DestinationDataDirectory -LogFileDirectory $DestinationLogDirectory -DestinationFileStreamDirectory $DestinationFileStreamDirectory -DatabaseFileSuffix $DestinationFileSuffix -DatabaseFilePrefix $DestinationFilePrefix -DatabaseNamePrefix $RestoredDatabaseNamePrefix -ReplaceDatabaseName $DatabaseName -Continue:$Continue -ReplaceDbNameInFile:$ReplaceDbNameInFile -FileMapping $FileMapping -PathSep $pathSep
+            $BackupHistory = $BackupHistory | Format-DbaBackupInformation -DataFileDirectory $DestinationDataDirectory -LogFileDirectory $DestinationLogDirectory -DestinationFileStreamDirectory $DestinationFileStreamDirectory -DatabaseFileSuffix $DestinationFileSuffix -DatabaseFilePrefix $DestinationFilePrefix -DatabaseNamePrefix $RestoredDatabaseNamePrefix -ReplaceDatabaseName $DatabaseName -Continue:$Continue -ReplaceDbNameInFile:$ReplaceDbNameInFile -FileMapping $FileMapping -PathSep $pathSep
```
Comment (Stuart-Moore) [inline functions/Restore-DbaDatabase.ps1:682]: Looks good to me

File: functions/Restore-DbaDatabase.ps1 [hunk 2]
```diff
@@ -588,7 +588,9 @@ function Restore-DbaDatabase {
                             }
                         }
                     }
-                    $BackupHistory += $F | Select-Object *, @{ Name = "ServerName"; Expression = { $_.SqlInstance } }, @{ Name = "BackupStartDate"; Expression = { $_.Start -as [DateTime] } }
+                    # Fix #5036 by implementing a deep copy of the FileList
+                    $f.FileList = $f.FileList | Select-Object *
+                    $BackupHistory += $f | Select-Object *, @{ Name = "ServerName"; Expression = { $_.SqlInstance } }, @{ Name = "BackupStartDate"; Expression = { $_.Start -as [DateTime] } }
```
Comment (Stuart-Moore) [inline functions/Restore-DbaDatabase.ps1:593]: This should do it.

File: tests/Restore-DbaDatabase.Tests.ps1 [hunk 3]
```diff
@@ -143,6 +143,16 @@ Describe "$CommandName Integration Tests" -Tag "IntegrationTests" {
         }
     }
 
+    Context "Replace databasename in Restored File, but don't change backup history #5036" {
```
Comment (Stuart-Moore) [inline tests/Restore-DbaDatabase.Tests.ps1:146]: Tests are good 👍 

 Should stop your hard work getting broken further on down the line

**Review by Stuart-Moore — APPROVED** [review]

**Comment by Stuart-Moore** [comment]
Happy with all of that. Merging in the new changes from dev to make sure all is still good, then i'll merge it in properly. If I get distracted, anyone else who wants to merge it, please do.

**Comment by Stuart-Moore** [comment]
Squashed and merged

Nice work @andreasjordan

**Comment by potatoqualitee** [comment]
Thank you all! 🙇🏼  I'll release this hopefully within the next 48 hours. I'm working on an auto releaser and happy to have something in the queue.

**Pull request merged**